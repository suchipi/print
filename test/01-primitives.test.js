const print = require("../print");

describe("Primitives", () => {
	it("prints null", () => {
		expect(print(null)).toMatchSnapshot();
	});
	it("prints undefined", () => {
		expect(print(undefined)).toMatchSnapshot();
	});

	describe("BigInts", () => {
		it("prints positive integers", () => {
			expect(print(42n)).toMatchSnapshot();
		});
		it("prints negative integers", () => {
			expect(print(-42n)).toMatchSnapshot();
		});
	});

	describe("Booleans", () => {
		it("prints true", () => {
			expect(print(true)).toMatchSnapshot();
		});
		it("prints false", () => {
			expect(print(false)).toMatchSnapshot();
		});
	});

	describe("Numbers", () => {
		it("prints positive integers", () => {
			expect(print(42)).toMatchSnapshot();
		});
		it("prints negative integers", () => {
			expect(print(-42)).toMatchSnapshot();
		});
		it("prints positive floats", () => {
			expect(print(4.2)).toMatchSnapshot();
		});
		it("prints negative floats", () => {
			expect(print(-4.2)).toMatchSnapshot();
		});
		it("prints positive infinity", () => {
			expect(print(Infinity)).toMatchSnapshot();
		});
		it("prints negative infinity", () => {
			expect(print(-Infinity)).toMatchSnapshot();
		});
		it("prints positive zero", () => {
			expect(print(0)).toMatchSnapshot();
		});
		it("prints negative zero", () => {
			expect(print(-0)).toMatchSnapshot();
		});
		it("shortens long numbers", () => {
			expect(print(1e64)).toMatchSnapshot();
		});
		it("prints NaN", () => {
			expect(print(NaN)).toMatchSnapshot();
		});

		it("identifies Math.* constants", () => {
			expect(print(Math.E)).toMatchSnapshot();
			expect(print(Math.LN10)).toMatchSnapshot();
			expect(print(Math.LN2)).toMatchSnapshot();
			expect(print(Math.LOG10E)).toMatchSnapshot();
			expect(print(Math.LOG2E)).toMatchSnapshot();
			expect(print(Math.PI)).toMatchSnapshot();
			expect(print(Math.SQRT1_2)).toMatchSnapshot();
			expect(print(Math.SQRT2)).toMatchSnapshot();
		});

		it("identifies Number.* constants", () => {
			expect(print(Number.EPSILON)).toMatchSnapshot();
			expect(print(Number.MIN_VALUE)).toMatchSnapshot();
			expect(print(Number.MAX_VALUE)).toMatchSnapshot();
			expect(print(Number.MIN_SAFE_INTEGER)).toMatchSnapshot();
			expect(print(Number.MAX_SAFE_INTEGER)).toMatchSnapshot();
		});
	});

	describe("Strings", () => {
		it("prints strings", () => {
			expect(print("Foo")).toMatchSnapshot();
		});
		it("escapes tabs", () => {
			expect(print("\t")).toMatchSnapshot();
		});
		it("escapes line-feeds", () => {
			expect(print("\n")).toMatchSnapshot();
		});
		it("escapes form-feeds", () => {
			expect(print("\f")).toMatchSnapshot();
		});
		it("escapes carriage-returns", () => {
			expect(print("\r")).toMatchSnapshot();
		});
		it("escapes vertical-tabs", () => {
			expect(print("\v")).toMatchSnapshot();
		});
		it("escapes null-bytes", () => {
			expect(print("\0")).toMatchSnapshot();
		});
		it("escapes backspaces", () => {
			expect(print("\b")).toMatchSnapshot();
		});
		it("escapes backslashes", () => {
			expect(print("\\")).toMatchSnapshot();
		});
		it("escapes bell characters", () => {
			expect(print("\x07")).toMatchSnapshot();
		});
		it("escapes ASCII escapes", () => {
			expect(print("\x1B")).toMatchSnapshot();
		});
		it("escapes other controls", () => {
			const c0 = [1, 2, 3, 4, 5, 6, 28, 29, 30, 31];
			const c1 = new Array(16).fill(0).map((x, i) => i + 128);
			for (let i = 14; i < 27; i++) {
				c0.push(i);
			}
			for (const code of [...c0, ...c1]) {
				const hex = code.toString(16).padStart(2, "0").toUpperCase();
				expect(print(String.fromCharCode(code))).toMatchSnapshot(hex);
			}
		});
	});

	describe("Symbols", () => {
		const desc = Object.getOwnPropertyDescriptor(globalThis, "String");
		afterEach(() => Object.defineProperty(globalThis, "String", desc));

		it("prints symbol values", () => {
			const foo = Symbol("Foo");
			expect(print(foo)).toMatchSnapshot();
		});

		it("identifies well-known symbols", () => {
			expect(print(Symbol.asyncIterator)).toMatchSnapshot();
			expect(print(Symbol.hasInstance)).toMatchSnapshot();
			expect(print(Symbol.isConcatSpreadable)).toMatchSnapshot();
			expect(print(Symbol.iterator)).toMatchSnapshot();
			expect(print(Symbol.match)).toMatchSnapshot();
			expect(print(Symbol.replace)).toMatchSnapshot();
			expect(print(Symbol.search)).toMatchSnapshot();
			expect(print(Symbol.species)).toMatchSnapshot();
			expect(print(Symbol.split)).toMatchSnapshot();
			expect(print(Symbol.toPrimitive)).toMatchSnapshot();
			expect(print(Symbol.toStringTag)).toMatchSnapshot();
			expect(print(Symbol.unscopables)).toMatchSnapshot();
		});

		it("doesn't identify them if `opts.noAmp` is enabled", () => {
			const { match } = Symbol;
			expect(print(match)).toMatchSnapshot();
			expect(print(match, { noAmp: false })).toMatchSnapshot();
			expect(print(match, { noAmp: true })).toMatchSnapshot();
		});

		it("doesn't confuse them with identically-named symbols", () => {
			expect(print(Symbol("Symbol.iterator"))).toMatchSnapshot();
			expect(print(Symbol("iterator"))).toMatchSnapshot();
			expect(print(Symbol.iterator)).toMatchSnapshot();
		});

		it("ensures value is always enclosed by `Symbol(â€¦)`", () => {
			Object.defineProperty(globalThis, "String", {
				configurable: true,
				value: (input) =>
					typeof input === "symbol"
						? input.toString().replace(/^Symbol\(|\)$/g, "")
						: desc.value.call(globalThis, input),
			});
			const sym = Symbol("Foo");
			expect(String(sym)).toEqual("Foo");
			expect(print(sym)).toMatchSnapshot();
		});
	});
});
