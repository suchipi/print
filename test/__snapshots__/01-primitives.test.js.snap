// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primitives BigInts prints negative integers 1`] = `-42n`;

exports[`Primitives BigInts prints positive integers 1`] = `42n`;

exports[`Primitives Booleans prints false 1`] = `false`;

exports[`Primitives Booleans prints true 1`] = `true`;

exports[`Primitives Numbers identifies Math.* constants 1`] = `Math.E`;

exports[`Primitives Numbers identifies Math.* constants 2`] = `Math.LN10`;

exports[`Primitives Numbers identifies Math.* constants 3`] = `Math.LN2`;

exports[`Primitives Numbers identifies Math.* constants 4`] = `Math.LOG10E`;

exports[`Primitives Numbers identifies Math.* constants 5`] = `Math.LOG2E`;

exports[`Primitives Numbers identifies Math.* constants 6`] = `Math.PI`;

exports[`Primitives Numbers identifies Math.* constants 7`] = `Math.SQRT1_2`;

exports[`Primitives Numbers identifies Math.* constants 8`] = `Math.SQRT2`;

exports[`Primitives Numbers identifies Number.* constants 1`] = `Number.EPSILON`;

exports[`Primitives Numbers identifies Number.* constants 2`] = `Number.MIN_VALUE`;

exports[`Primitives Numbers identifies Number.* constants 3`] = `Number.MAX_VALUE`;

exports[`Primitives Numbers identifies Number.* constants 4`] = `Number.MIN_SAFE_INTEGER`;

exports[`Primitives Numbers identifies Number.* constants 5`] = `Number.MAX_SAFE_INTEGER`;

exports[`Primitives Numbers prints NaN 1`] = `NaN`;

exports[`Primitives Numbers prints negative floats 1`] = `-4.2`;

exports[`Primitives Numbers prints negative infinity 1`] = `-Infinity`;

exports[`Primitives Numbers prints negative integers 1`] = `-42`;

exports[`Primitives Numbers prints negative zero 1`] = `-0`;

exports[`Primitives Numbers prints positive floats 1`] = `4.2`;

exports[`Primitives Numbers prints positive infinity 1`] = `Infinity`;

exports[`Primitives Numbers prints positive integers 1`] = `42`;

exports[`Primitives Numbers prints positive zero 1`] = `0`;

exports[`Primitives Numbers shortens long numbers 1`] = `1e+64`;

exports[`Primitives Strings escapes ASCII escapes 1`] = `"\\e"`;

exports[`Primitives Strings escapes backslashes 1`] = `"\\\\"`;

exports[`Primitives Strings escapes backspaces 1`] = `"\\b"`;

exports[`Primitives Strings escapes bell characters 1`] = `"\\a"`;

exports[`Primitives Strings escapes carriage-returns 1`] = `"\\r"`;

exports[`Primitives Strings escapes form-feeds 1`] = `"\\f"`;

exports[`Primitives Strings escapes line-feeds 1`] = `"\\n"`;

exports[`Primitives Strings escapes null-bytes 1`] = `"\\0"`;

exports[`Primitives Strings escapes other controls: 01 1`] = `"\\x01"`;

exports[`Primitives Strings escapes other controls: 02 1`] = `"\\x02"`;

exports[`Primitives Strings escapes other controls: 03 1`] = `"\\x03"`;

exports[`Primitives Strings escapes other controls: 04 1`] = `"\\x04"`;

exports[`Primitives Strings escapes other controls: 05 1`] = `"\\x05"`;

exports[`Primitives Strings escapes other controls: 06 1`] = `"\\x06"`;

exports[`Primitives Strings escapes other controls: 0E 1`] = `"\\x0E"`;

exports[`Primitives Strings escapes other controls: 0F 1`] = `"\\x0F"`;

exports[`Primitives Strings escapes other controls: 1A 1`] = `"\\x1A"`;

exports[`Primitives Strings escapes other controls: 1C 1`] = `"\\x1C"`;

exports[`Primitives Strings escapes other controls: 1D 1`] = `"\\x1D"`;

exports[`Primitives Strings escapes other controls: 1E 1`] = `"\\x1E"`;

exports[`Primitives Strings escapes other controls: 1F 1`] = `"\\x1F"`;

exports[`Primitives Strings escapes other controls: 8A 1`] = `"\\x8A"`;

exports[`Primitives Strings escapes other controls: 8B 1`] = `"\\x8B"`;

exports[`Primitives Strings escapes other controls: 8C 1`] = `"\\x8C"`;

exports[`Primitives Strings escapes other controls: 8D 1`] = `"\\x8D"`;

exports[`Primitives Strings escapes other controls: 8E 1`] = `"\\x8E"`;

exports[`Primitives Strings escapes other controls: 8F 1`] = `"\\x8F"`;

exports[`Primitives Strings escapes other controls: 10 1`] = `"\\x10"`;

exports[`Primitives Strings escapes other controls: 11 1`] = `"\\x11"`;

exports[`Primitives Strings escapes other controls: 12 1`] = `"\\x12"`;

exports[`Primitives Strings escapes other controls: 13 1`] = `"\\x13"`;

exports[`Primitives Strings escapes other controls: 14 1`] = `"\\x14"`;

exports[`Primitives Strings escapes other controls: 15 1`] = `"\\x15"`;

exports[`Primitives Strings escapes other controls: 16 1`] = `"\\x16"`;

exports[`Primitives Strings escapes other controls: 17 1`] = `"\\x17"`;

exports[`Primitives Strings escapes other controls: 18 1`] = `"\\x18"`;

exports[`Primitives Strings escapes other controls: 19 1`] = `"\\x19"`;

exports[`Primitives Strings escapes other controls: 80 1`] = `"\\x80"`;

exports[`Primitives Strings escapes other controls: 81 1`] = `"\\x81"`;

exports[`Primitives Strings escapes other controls: 82 1`] = `"\\x82"`;

exports[`Primitives Strings escapes other controls: 83 1`] = `"\\x83"`;

exports[`Primitives Strings escapes other controls: 84 1`] = `"\\x84"`;

exports[`Primitives Strings escapes other controls: 85 1`] = `"\\x85"`;

exports[`Primitives Strings escapes other controls: 86 1`] = `"\\x86"`;

exports[`Primitives Strings escapes other controls: 87 1`] = `"\\x87"`;

exports[`Primitives Strings escapes other controls: 88 1`] = `"\\x88"`;

exports[`Primitives Strings escapes other controls: 89 1`] = `"\\x89"`;

exports[`Primitives Strings escapes tabs 1`] = `"\\t"`;

exports[`Primitives Strings escapes vertical-tabs 1`] = `"\\v"`;

exports[`Primitives Strings prints strings 1`] = `"Foo"`;

exports[`Primitives Symbols doesn't confuse them with identically-named symbols 1`] = `Symbol(Symbol.iterator)`;

exports[`Primitives Symbols doesn't confuse them with identically-named symbols 2`] = `Symbol(iterator)`;

exports[`Primitives Symbols doesn't confuse them with identically-named symbols 3`] = `@@iterator`;

exports[`Primitives Symbols doesn't identify them if \`opts.noAmp\` is enabled 1`] = `@@match`;

exports[`Primitives Symbols doesn't identify them if \`opts.noAmp\` is enabled 2`] = `@@match`;

exports[`Primitives Symbols doesn't identify them if \`opts.noAmp\` is enabled 3`] = `Symbol(Symbol.match)`;

exports[`Primitives Symbols ensures value is always enclosed by \`Symbol(â€¦)\` 1`] = `Symbol(Foo)`;

exports[`Primitives Symbols identifies well-known symbols 1`] = `@@asyncIterator`;

exports[`Primitives Symbols identifies well-known symbols 2`] = `@@hasInstance`;

exports[`Primitives Symbols identifies well-known symbols 3`] = `@@isConcatSpreadable`;

exports[`Primitives Symbols identifies well-known symbols 4`] = `@@iterator`;

exports[`Primitives Symbols identifies well-known symbols 5`] = `@@match`;

exports[`Primitives Symbols identifies well-known symbols 6`] = `@@replace`;

exports[`Primitives Symbols identifies well-known symbols 7`] = `@@search`;

exports[`Primitives Symbols identifies well-known symbols 8`] = `@@species`;

exports[`Primitives Symbols identifies well-known symbols 9`] = `@@split`;

exports[`Primitives Symbols identifies well-known symbols 10`] = `@@toPrimitive`;

exports[`Primitives Symbols identifies well-known symbols 11`] = `@@toStringTag`;

exports[`Primitives Symbols identifies well-known symbols 12`] = `@@unscopables`;

exports[`Primitives Symbols prints symbol values 1`] = `Symbol(Foo)`;

exports[`Primitives prints null 1`] = `null`;

exports[`Primitives prints undefined 1`] = `undefined`;
